File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
def sorted_array_to_bst(nums)
    return nil if nums.empty?
    build_bst(nums, 0, nums.length - 1)
end

def build_bst(nums, start, end_idx)
    return nil if start > end_idx
    
    mid = start + (end_idx - start) / 2
    node = TreeNode.new(nums[mid])
    node.left = build_bst(nums, start, mid - 1)
    node.right = build_bst(nums, mid + 1, end_idx)
    
    node
end

File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function sortedArrayToBST(nums: number[]): TreeNode | null {
    return buildBST(nums, 0, nums.length - 1);
}

function buildBST(nums: number[], start: number, end: number): TreeNode | null {
    if (start > end) return null;
    
    const mid = start + Math.floor((end - start) / 2);
    const node = new TreeNode(nums[mid]);
    node.left = buildBST(nums, start, mid - 1);
    node.right = buildBST(nums, mid + 1, end);
    
    return node;
}